controllers/workspace/finalize.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/devfile/devworkspace-operator/pkg/common"
+	"github.com/devfile/devworkspace-operator/pkg/config"
+	"github.com/go-logr/logr"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	pvcClaimMountPath   = "/tmp/devworkspaces/"
+	cleanupCommandFmt   = "rm -rf %s"
+	pvcCleanupFinalizer = "storage.controller.devfile.io"
+)
+
+var (
+	cleanupJobCompletions  = int32(1)
+	cleanupJobBackoffLimit = int32(3)
+)
+
+func (r *DevWorkspaceReconciler) setFinalizer(ctx context.Context, workspace *v1alpha2.DevWorkspace) (ok bool, err error) {
+	if !isFinalizerNecessary(workspace) || hasFinalizer(workspace) {
+		return true, nil
+	}
+	workspace.SetFinalizers(append(workspace.Finalizers, pvcCleanupFinalizer))
+	return false, r.Update(ctx, workspace)
+}
+
+func (r *DevWorkspaceReconciler) finalize(ctx context.Context, log logr.Logger, workspace *v1alpha2.DevWorkspace) error {
+	if !hasFinalizer(workspace) {
+		return nil
+	}
+	specJob, err := r.getSpecCleanupJob(workspace)
+	if err != nil {
+		return err
+	}
+	clusterJob, err := r.getClusterCleanupJob(ctx, workspace)
+	if err != nil {
+		return err
+	}
+	if clusterJob == nil {
+		err := r.Create(ctx, specJob)
+		if err != nil && !k8sErrors.IsAlreadyExists(err) {
+			return err
+		}
+		return nil
+	}
+	if !equality.Semantic.DeepDerivative(specJob.Spec, clusterJob.Spec) {
+		err := r.Patch(ctx, specJob, client.Merge)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	for _, condition := range clusterJob.Status.Conditions {
+		if condition.Status != corev1.ConditionTrue {
+			continue
+		}
+		switch condition.Type {
+		case batchv1.JobComplete:
+			log.Info("PVC clean up job successful; clearing finalizer")
+			clearFinalizer(workspace)
+			return r.Update(ctx, workspace)
+		case batchv1.JobFailed:
+			log.Error(fmt.Errorf("PVC clean up job failed: message %q", condition.Message),
+				"Failed to clean PVC on workspace deletion")
+			clearFinalizer(workspace)
+			return r.Update(ctx, workspace)
+		}
+	}
+	return nil
+}
+
+func (r *DevWorkspaceReconciler) getSpecCleanupJob(workspace *v1alpha2.DevWorkspace) (*batchv1.Job, error) {
+	workspaceId := workspace.Status.WorkspaceId
+	pvcName := config.ControllerCfg.GetWorkspacePVCName()
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      common.PVCCleanupJobName(workspaceId),
+			Namespace: workspace.Namespace,
+			Labels: map[string]string{
+				config.WorkspaceIDLabel: workspaceId,
+			},
+		},
+		Spec: batchv1.JobSpec{
+			Completions:  &cleanupJobCompletions,
+			BackoffLimit: &cleanupJobBackoffLimit,
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
```

Comment:
Is not it necessary to reuse the same security context as we use for deployment workspace deployment https://github.com/devfile/devworkspace-operator/blob/master/controllers/workspace/provision/deployment.go#L137

Or missing security context here leads to using root to avoid permissions issues. If so - it would be better to set it explicitely

Reply:
I added the security context in, and then also realized that we were setting it somewhat incorrectly -- GID should be 0 (at least, that's what most of our images assume).